### DISCUSSION TITLE
Configuration of a private ECR Registry with a token for Renovate

### DISCUSSION THREAD

**Discussion by Johannes1509** [renovatebot/renovate#18755] [Request Help]
Hi I am trying to connect Renovate to a private AWS ECR so that renovate can check for new dependencies for images from there. However, the connection does not work, in the log I find the following error:
```
WARN: Error obtaining docker token (repository=docker/image-test, baseBranch=admin_repo_test)
       "registryHost": "https://996594000000.dkr.ecr.eu-central-1.amazonaws.com",
       "dockerRepository": "cert-manager-controller",
       "err": {
         "name": "RequestError",
         "code": "EPROTO",
         "timings": {
           "start": 1667551481069,
           "socket": 1667551481124,
           "lookup": 1667551481124,
           "connect": 1667551481124,
           "upload": 1667551481130,
           "error": 1667551481130,
           "phases": {"wait": 55, "dns": 0, "tcp": 0, "request": 6, "total": 61}
         },
         "message": "write EPROTO 140586387842944:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:332:\n",
         "stack": "RequestError: write EPROTO 140586387842944:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:332:\n\n    at ClientRequest.<anonymous> (/usr/src/app/node_modules/got/dist/source/core/index.js:970:111)\n    at Object.onceWrapper (events.js:520:26)\n    at ClientRequest.emit (events.js:412:35)\n    at ClientRequest.emit (domain.js:475:12)\n    at ClientRequest.origin.emit (/usr/src/app/node_modules/@szmarczak/http-timer/dist/source/index.js:43:20)\n    at TLSSocket.socketErrorListener (_http_client.js:475:9)\n    at TLSSocket.emit (events.js:412:35)\n    at TLSSocket.emit (domain.js:475:12)\n    at emitErrorNT (internal/streams/destroy.js:106:8)\n    at emitErrorCloseNT (internal/streams/destroy.js:74:3)\n    at processTicksAndRejections (internal/process/task_queues.js:82:21)\n    at WriteWrap.onWriteComplete [as oncomplete] (internal/stream_base_commons.js:94:16)",
         "options": {
           "headers": {
             "user-agent": "RenovateBot/32.236.3 (https://github.com/renovatebot/renovate)",
             "accept": "application/json",
             "accept-encoding": "gzip, deflate, br"
           },
           "url": "https://996594000000.dkr.ecr.eu-central-1.amazonaws.com/v2/cert-manager-controller/tags/list?n=1000",
           "hostType": "docker",
           "username": "",
           "password": "",
           "method": "GET",
           "http2": false
         }
       }
```

My configuration:
config.js:
```
module.exports = {
    platform: 'gitlab',
    endpoint: 'https://git.undbt.de/api/v4/',
    token: process.env.RENOVATETOKEN,
    logFileLevel: 'debug',
    logLevel: 'debug',
    logFile: 'renovate.log',
    onboarding: true,
     onboardingConfig: {
        extends: [
            'local>renovate-imageupdate'
        ],
    },

    repositories: [
        "docker/image-test",
    ],
    commitMessagePrefix: "[renovate bot] ",
    baseBranches: ["admin_repo_test"],
    enabled: true,
    separateMajorMinor: true,
    separateMinorPatch: true,
    automerge: false,
    major: {
        enabled: false,
        automerge: false
    },
    minor: {
        enable: false,
        automerge: false
    },
    patch: {
        enable: true,
        automerge: false
    },
    "packageRules": [{
        "registryUrls": ["https://996594000000.dkr.ecr.eu-central-1.amazonaws.com"]
    }],
    hostRules: [
        {
            "hostType": "docker",
            "hostName": "https://996594000000.dkr.ecr.eu-central-1.amazonaws.com",
            "username": "AWS",
            "password": process.env.ECRTOKEN
        }
    ]
};
```

`process.env.ECRTOKEN` is a valid token which I retrieve via `aws ecr get-login-password` in the Gitlab CI pipeline where renovate is also executed. Does anyone have an idea what the error means?

What I also noticed: Despite specifying username and password, neither appears in the debug log (see above) where username and password is empty.

**Comment by viceice** [comment]
which version of renovate are you using?

`logLevel: 'debug',` is not a valid option

https://docs.renovatebot.com/troubleshooting/#self-hosted

**Reply by Johannes1509** [reply]
Thx for the loglevel hint ;) Im using renovate 34.17.0 as a docker image in a gitlab ci pipeline.

**Reply by viceice** [reply]
you need to configure accesskey and secretkey

https://github.com/renovatebot/renovate/blob/4858ab059d5774913979933adc6afe95d850f8ee/lib/modules/datasource/docker/index.ts#L237-L244

**Reply by Johannes1509** [reply]
Mmmh, I'm using username and password with credentials as metioned here: https://github.com/renovatebot/renovate/issues/3968. If I'm opening the failing url (https://996594000000.dkr.ecr.eu-central-1.amazonaws.com/v2/cert-manager-controller/tags/list?n=1000) in my browser and enter the credentials in the login prompt as in my config.js I'm receiving a valid response from the ecr.
As provided by the offical aws cli docs the username "AWS" + the ecr-cli token is a valid username/password combination for docker login https://docs.aws.amazon.com/cli/latest/reference/ecr/get-login-password.html

**Reply by viceice** [reply]
as mentioned in #3968 you need access and secret key. renovate will call the aws api to get a ecr token internally

https://github.com/renovatebot/renovate/blob/4858ab059d5774913979933adc6afe95d850f8ee/lib/modules/datasource/docker/index.ts#L246-L252

**Reply by Johannes1509** [reply]
Is there any way to use a already generated ecr token for renovate login to not have to put secret access key information in the config file?

**Reply by viceice** [reply]
no, you can also pass the secrets via ENV, like you do for the token

**Reply by lukasmrtvy** [reply]
@viceice You can use this in GH workflow: ( "credential-less" AWS and ECR auth )
```
    steps:
      - uses: actions/checkout@v3
      - uses: aws-actions/configure-aws-credentials@v1
        with:
          aws-region: us-east-1
          role-to-assume: arn:aws:iam::1111111111111:role/my-role-with-ecr-permission
      - uses: aws-actions/amazon-ecr-login@v1
      - run: aws sts get-caller-identity
      - uses: renovatebot/github-action@v32.238.4
        ...
```
`aws-sdk-js` should handle both use cases, there is no point specifying additional "static" credentials. This should work for docker and aws-machine-image data sources and all other AWS data sources in the future.

**Reply by viceice** [reply]
i don't think that it works as the renovate GitHub action doesn't pass the AWS ENV values to the renovate container.

**Reply by lukasmrtvy** [reply]
https://github.com/renovatebot/github-action/issues/585

**Comment by jheddings** [comment]
Renovate requires AWS credentials for authentication against private ECR repos.  There is a regex that ignores `username` and `password` for basic auth:

https://github.com/renovatebot/renovate/blob/bde951257a79737351859d1c08db0ad214e61494/lib/modules/datasource/docker/index.ts#L105-L115

I've started a PR to support using the login password (as well as authorization token) from the CLI.  Would appreciate your thoughts there also - https://github.com/renovatebot/renovate/discussions/20326
